toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -165,91 +65,210 @@ static auto GetBinaryOp(TokenKind kind) -> Semantics::BinaryOperator::Op {
   }
 }
 
-void SemanticsIRFactory::TransformInfixOperator(
-    llvm::SmallVector<Semantics::NodeRef, 0>& nodes, ParseTree::Node node,
-    Semantics::NodeId target_id) {
-  CARBON_CHECK(parse_tree().node_kind(node) == ParseNodeKind::InfixOperator());
-
-  auto token = parse_tree().node_token(node);
-  auto token_kind = tokens_->GetKind(token);
-  auto op = GetBinaryOp(token_kind);
-
-  auto rhs_id = next_id();
-  auto lhs_id = next_id();
-  nodes.push_back(semantics_.nodes_.Store(
-      Semantics::BinaryOperator(node, target_id, op, lhs_id, rhs_id)));
-  auto subtree = ParseSubtreeConsumer::ForParent(parse_tree(), node);
-  TransformExpression(nodes, subtree.RequireConsume(), rhs_id);
-  TransformExpression(nodes, subtree.RequireConsume(), lhs_id);
-}
-
-/*
-auto SemanticsIRFactory::TransformParameterList(ParseTree::Node node)
-    -> llvm::SmallVector<Semantics::PatternBinding, 0> {
-  CARBON_CHECK(parse_tree().node_kind(node) ==
-ParseNodeKind::ParameterList());
-
-  auto subtree = ParseSubtreeConsumer::ForParent(parse_tree(), node);
-  RequireNodeEmpty(subtree.RequireConsume(ParseNodeKind::ParameterListEnd()));
-
-  llvm::SmallVector<Semantics::PatternBinding, 0> params;
-  if (auto first_param_node =
-          subtree.TryConsume(ParseNodeKind::PatternBinding())) {
-    params.push_back(TransformPatternBinding(*first_param_node));
-
-    while (auto comma_node =
-               subtree.TryConsume(ParseNodeKind::ParameterListComma())) {
-      RequireNodeEmpty(*comma_node);
-      params.push_back(TransformPatternBinding(
-          subtree.RequireConsume(ParseNodeKind::PatternBinding())));
+// bazel run //toolchain/driver:carbon dump semantics-ir
+// testdata/function/basic.carbon
+void SemanticsIRFactory::Build() {
+  // Silence "unused" build warning.
+  llvm::SmallVector<TraversalStackEntry, 0> node_stack;
+  auto range = parse_tree().postorder();
+  for (auto it = range.begin();; ++it) {
+    auto parse_node = *it;
+    int32_t subtree_size = parse_tree().node_subtree_size(parse_node);
+    switch (auto parse_kind = parse_tree().node_kind(parse_node)) {
+      case ParseNodeKind::CodeBlock(): {
+        // Just merges children.
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+        while (subtree_size > 1) {
+          subtree_size -=
+              parse_tree().node_subtree_size(node_stack.back().parse_node);
+          sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                        node_stack.back().sem_ir.end());
```

Comment:
This looks to be quadratic in the number of statements in the code block to me. Could we build a list of lists / ranges, reverse it, then concatenate? Or first compute the number of entries from `node_stack`, and then walk them in reverse order and concatenate their IR, and then pop them all at once?

If you're just going for correctness for now and want to defer this for a future change, I think that's fine, but we should at least have a TODO so we don't forget.

Reply:
Hm, if we make all `TraversalStackEntry`s share the same vector of `NodeRef`s as I mentioned above, I think this can even be done with no copies, by setting the new stack entry's first owned index to the start index of the first substatement and setting its last owned index to the end index of the last substatement.

Reply:
I'm not happy with this -- to me, this is part of the TODO to rethink the sem_ir approach. But it helps with test consistency for now.